src/coreclr/jit/simdcodegenxarch.cpp

```diff
@@ -446,7 +446,7 @@ void CodeGen::genSIMDScalarMove(
                 }
                 else
                 {
-                    genSIMDZero(targetType, TYP_FLOAT, targetReg);
```

Comment:
I'm slightly worried about how this might impact CSE, particularly since (AFAIR) it relies on having the handle available (which requires the baseType to resolve)

Reply:
These changes are (and I think the intent for them is to stay that way) all contained to codegen, so this shouldn't be a problem?

Reply:
It might be fine for this case, but it does lose the ability to generate different instructions based on type (which we somewhat do for hwintrinsics).

I guess a bit of it will be moot when I eventually port this over to the `SimdAsHWIntrinsic` path...

Reply:
> It might be fine for this case, but it does lose the ability to generate different instructions based on type (which we somewhat do for hwintrinsics).

My understanding is that some intrinsic generate different code depending on their base type, for example, init(SIMD16, 1) will need to know the base type to know how many elements are in the SIMD16.
The other intrinsics in some cases do not depend on base type, like `SIMDIntrinsicInit(0)`, so we should not include baseType in CSE/VN considerations.

> It might be fine for this case, but it does lose the ability to generate different instructions based on type (which we somewhat do for hwintrinsics).

I don't see what is lost here, the generated code does not depend on the base type.

Reply:
There are multiple instructions for zeroing and while they are treated the same on most modern CPUs, not all CPUs do.

And so, it may be that certain hardware we want to generate `pxor` or `xorpd` instead of `xorps`. Likewise, if we ever add AVX-512 or SVE support, we may want different logic here for various masking or other operations that can occur.
and so carrying the type through to codegen is goodness IMO, even if today we happen to ignore it.

Reply:
> And so, it may be that certain hardware we want to generate pxor or xorpd instead of xorps. Likewise, if we ever add AVX-512 or SVE support, we may want different logic here for various masking or other operations that can occur.

It could be that we want to generate different instruction on different hardware, but will it be possible that the optimal zeroing for a given size would be different? Could you give me an example when the optimal instruction to generate `Vector256<int>(0)` would be different from `Vector256<float>` or `Vector256<double>` etc. In my understanding, the optimal instruction to use here depends only on target platform and SIMD size, not on the base type of the simd.

I understand that for other intrinsics it could be different, but my change only affects zero-init.

Reply:
For both `TYP_SIMD16` and `TYP_SIMD32` there are the following "simd zeroing" instructions:
* `xorps` - `float`
* `xorpd` - `double`
* `pxor` - `integer`
Where `xmm` is used for `TYP_SIMD16` and `ymm` is used for `TYP_SIMD32`

On modern Intel/AMD (the architecture manuals list Silvermont/Ryzen and newer), all three of these are dependency breaking/zeroing idioms and should actually be replaced in the renamer and so *should* be equivalent/interchangeable.

However, on some hardware (particularly "older" hardware, such as Intel Core microarchitecture), the recommendation is to not mix floating-point and integer code:
> 3.5.1.10 Mixing SIMD Data Type
> Previous microarchitectures (before Intel Core microarchitecture) do not have explicit restrictions on mixing integer and floating-point (FP) operations on XMM registers. For Intel Core microarchitecture, mixing integer and floating-point operations on the content of an XMM register can degrade performance. Software should avoid mixed-use of integer/FP operation on XMM registers. Specifically:
• Use SIMD integer operations to feed SIMD integer operations. Use PXOR for idiom.
• Use SIMD floating-point operations to feed SIMD floating-point operations. Use XORPS for idiom.
• When floating-point operations are bitwise equivalent, use PS data type instead of PD data type. 
MOVAPS and MOVAPD do the same thing, but MOVAPS takes one less byte to encode the instruction.

Likewise, once you get to AVX-512, there are additional instructions such as `pxord` and `pxorq`, which get used instead because it is impactful for masked instructions.

Reply:
I see, thanks for the explanation, so would you prefer to lie about the type when it is not available to better support a future possibility of an arch where `void genSIMDZero(var_types targetType, var_types baseType, regNumber targetReg);` will use `base_type`?

Reply:
I'd prefer we always pass the "correct" type for the operation here and to finish plumbing through things like `GenTreeJitIntrinsic->m_classLayout`, but short of that the existing behavior of lying works well enough today so we can still generate something.

Reply:
I guess I did not clearly describe the scenario that I want to support:
```
struct RandomStructWith16ByteSize
{
  int a;
  float b;
  int c;
  float d;
}
```
and we do:
```
RandomStructWith16ByteSize a = new RandomStructWith16ByteSize(); // zero init
```
if `a` can be put into an xmm register I want it to be there and generate a zero-init when there is no existing "correct type".

Reply:
Maybe we should just explicitly support `TYP_UNDEF` and we can handle it as appropriate (which is currently treating it as `TYP_FLOAT`) in codegen ?

For context, I've been slowly doing work to get rid of the older SIMD support (like `SimdIntrinsicInit`) and replacing it with the newer HWIntrinsic support. When that happens, `genSimdZero` will be replaced with `NI_Vector64_get_Zero`, `NI_Vector128_get_Zero`, and `NI_Vector256_get_zero` so everything can be centrally handled. -- #35421 and #37882 are the two past PRs and https://github.com/dotnet/runtime/pull/52288 is the latest

At that point, the codegen will generate different instructions based on the type by default and we'll need to handle `TYP_UNDEF` somehow anyways (different HWIntrinsics have different contracts for what instruction they'll generate, since they are mostly a 1-to-1 mapping with the hardware).